src/sharing/ShareWrapper.js

```diff
@@ -0,0 +1,257 @@
+  wrapper: {
+    flex: 1,
+    padding: 10,
+  },
+  container: {
+    flex: 1,
+  },
+  imagePreview: {
+    margin: 10,
+    borderRadius: 5,
+    width: 200,
+    height: 200,
+  },
+  actions: {
+    flexDirection: 'row',
+  },
+  button: {
+    flex: 1,
+    margin: 8,
+  },
+  previewText: {
+    color: 'white',
+    textAlign: 'center',
+    backgroundColor: '#000000a0',
+    position: 'absolute',
+    bottom: 0,
+    left: 0,
+    right: 0,
+  },
+});
+
+type SendStream = {|
+  stream: string,
+  topic: string,
+  type: 'stream',
+|};
+
+type SendPm = {|
+  selectedRecipients: $ReadOnlyArray<UserId>,
+  type: 'pm',
+|};
+
+type Props = $ReadOnly<{|
+  children: React$Node,
+  dispatch: Dispatch,
+
+  auth: Auth,
+  isSendButtonEnabled: (message: string) => boolean,
+  ownUserId: UserId,
+  sendTo: SendStream | SendPm,
+  sharedData: SharedData,
+|}>;
+
+type State = $ReadOnly<{|
+  message: string,
+  sending: boolean,
+  files: File[],
+|}>;
+
+/**
+ * Wraps Around different sharing screens,
+ * for minimal duplication of code.
+ */
+class ShareWrapper extends React.Component<Props, State> {
+  static contextType = TranslationContext;
+  context: GetText;
+
+  state = (() => {
+    const { sharedData } = this.props;
+    return {
+      message: sharedData.type === 'text' ? sharedData.sharedText : '',
+      sending: false,
+      files: sharedData.type === 'text' ? [] : sharedData.files,
+    };
+  })();
+
+  handleMessageChange = (message: string) => {
+    this.setState({ message });
+  };
+
+  setSending = () => {
+    this.setState({ sending: true });
+  };
+
+  /**
+   * Send received shared data as a message.
+   *
+   * Sends the shared data received from a 3rd party app and possibly modified
+   * by the user, to the server, as a message.
+   */
+  handleSend = async () => {
+    const _ = this.context;
+    const { auth, sendTo, sharedData } = this.props;
+    let messageToSend = this.state.message;
+
+    this.setSending();
+    showToast(_('Sending Message...'));
+    if (sharedData.type === 'file') {
+      const { files } = this.state;
+      for (let i = 0; i < files.length; i++) {
+        const response = await api.uploadFile(auth, files[i].url, files[i].name);
+        messageToSend += `\n[${files[i].name}](${response.uri})\n`;
+      }
+    }
+    const messageData =
+      sendTo.type === 'pm'
+        ? {
+            content: messageToSend,
+            type: 'private',
+            to: JSON.stringify(sendTo.selectedRecipients),
+          }
+        : {
+            content: messageToSend,
+            type: 'stream',
+            subject: sendTo.topic,
+            to: sendTo.stream,
+          };
+
+    try {
+      await api.sendMessage(auth, messageData);
+    } catch (err) {
+      showToast(_('Failed to send message'));
+      logging.error(err);
+      this.onShareTerminated();
+      return;
+    }
+    showToast(_('Message sent'));
+    this.onShareSuccess();
+  };
+
+  onShareTerminated = () => {
```

Comment:
"Terminated" is kind of ambiguous -- it can mean the thing got aborted because it wasn't working; or it can just mean that the thing ended one way or another, succeed or fail.

Here, what we mean is that it ended and wasn't a success -- this is something we won't call in the cases where we call `onShareSuccess`. So "fail" is a good word to contrast with "succeed".

How about `onShareFailure`?

Reply:
Oh I see, and we also call this if the user cancels. So "failure" is too specific.

"Cancel" is a good word, though -- in particular it can cover the case where something failed, and then our code cancelled the send in response to that. So how about `onShareCancelled`?